docs/dxp/7.x/en/installation-and-upgrades/setting-up-liferay-dxp/tuning-your-jvm.md

```diff
@@ -0,0 +1,103 @@
+# Tuning Your JVM
+
+Java Virtual Machine (JVM) tuning primarily focuses on adjusting the garbage collector and the Java memory heap. We used Oracle's 1.8 JVM for the reference architecture. You may choose other supported JVM versions and implementations. Please consult the  [Liferay DXP Compatibility Matrix](https://web.liferay.com/group/customer/dxp/support/compatibility-matrix) for additional compatible JVMs.
+
+Here are the JVM tuning topics:
+
+* [Garbage Collector](#garbage-collector)
+* [Code Cache](#code-cache)
+* [Java Heap](#java-heap)
+* [JVM Advanced Options](#jvm-advanced-options)
+
+Garbage collection is first.
+
+## Garbage Collector
+
+Choosing the appropriate garbage collector (GC) helps improve the responsiveness of your Liferay DXP deployment. Use the concurrent low pause collectors:
+
+```bash
+-XX:+UseParNewGC -XX:ParallelGCThreads=16 -XX:+UseConcMarkSweepGC
+-XX:+CMSParallelRemarkEnabled -XX:+CMSCompactWhenClearAllSoftRefs
+-XX:CMSInitiatingOccupancyFraction=85 -XX:+CMSScavengeBeforeRemark
+```
+
+You may choose from other available GC algorithms, including parallel throughput collectors and G1 collectors. Start tuning using parallel collectors in the new generation and concurrent mark sweep (CMS) in the old generation.
+
+**Note:** the `ParallelGCThreads` value (e.g., `ParallelGCThreads=16`) varies  based on the type of CPUs available. Set the value according to CPU specification. On Linux machines, report the number of available CPUs by running `cat /proc/cpuinfo`.
+
+**Note:** There are additional "new" algorithms like G1, but Liferay Engineering's tests for G1 indicated that it does not improve performance. Since your application performance may vary, you should add G1 to your testing and tuning plans.
+
+## Code Cache
+
+Java's just-in-time (JIT) compiler generates native code to improve performance. The default size is `48m`. This may not be sufficient for larger applications. Too small a code cache reduces performance, as the JIT isn't able to optimize high frequency methods. For Liferay DXP,  start with `64m` for the initial code cache size.
+
+```bash
+-XX:InitialCodeCacheSize=64m -XX:ReservedCodeCacheSize=96m
+```
+
+Examine the efficacy of the parameter changes by adding the following logging parameters:
+
+```bash
+-XX:+PrintCodeCache -XX:+PrintCodeCacheOnCompilation
+```
+
+## Java Heap
+
+When most people think about tuning the Java memory heap, they think of setting the maximum and minimum memory of the heap. Unfortunately, most deployments require far more sophisticated heap tuning to obtain optimal performance, including tuning the young generation size, tenuring durations, survivor spaces, and many other JVM internals.
+
+For most systems, it's best to start with at least the following memory settings:
+
+```bash
+server -XX:NewSize=700m -XX:MaxNewSize=700m -Xms2560m -Xmx2560m -XX:MetaspaceSize=512m
+-XX:MaxMetaspaceSize=512m -XX:SurvivorRatio=6 -XX:TargetSurvivorRatio=9 -XX:MaxTenuringThreshold=15
```

Comment:
@jhinkey `-XX:SurvivorRatio=6 -XX:TargetSurvivorRatio=9 -XX:MaxTenuringThreshold=15`

We've been using `-XX:SurvivorRatio=16 -XX:TargetSurvivorRatio=16 -XX:MaxTenuringThreshold=15` in recent 1~2 years, enabling survivor significantly reduces old gen GC pressure. Disabling survivor was never recommended for production, it was more of a special config for testing purpose.

Reply:
@dantewang I'm suspicious of the same value being used for `-XX:SurvivorRatio` and `-XX:TargetSurvivorRatio` because the first supposed to be a ratio of eden/survivor and the second supposed to the percentage of survivor space used after scavenge. I wouldn't expect those values to be the same. Are you sure?

Reply:
@jhinkey 
Firstly I mentioned benchmark settings because I thought the values in the original doc were both 0 (actually 6 and 9) :P

For benchmark settings, I think they just happened to be same in our config. Looking at the VisualGC picture from Login test case below, `-XX:SurvivorRatio=16` makes each survivor takes **1/18** of the specified new size (9g), which is 512m; and the eden space itself actually takes 8g.

`-XX:TargetSurvivorRatio=16` makes less objects being kept in survivor (more being promoted to old gen). I don't remember  why it was set to this value, but setting a smaller value helps mitigate spike in eden space promotion, which may happen in our benchmark in other test cases.

They are also dependent on actual use cases. The settings work well in our benchmark because we reduce object lifetime as much as possible when optimizing portal. But customer code may be different.

For customer production I think you can suggest customer to use the default value (50%) as a starting point of `-XX:TargetSurvivorRatio`. Anyway `9` is not good.

![Visual GC1596424798840](https://user-images.githubusercontent.com/1185063/100042414-5b57c800-2e46-11eb-8224-51266a8c4fa3.jpg)